packages/world/src/SystemCall.sol

```diff
@@ -0,0 +1,103 @@
+
+library SystemCall {
+  using ResourceSelector for bytes32;
+
+  /**
+   * Call the system at the given namespace.
+   * If the system is not public, the caller must have access to the namespace or name.
+   * Returns the success status and any return data.
+   */
+  function call(
+    address caller,
+    bytes32 resourceSelector,
+    bytes memory funcSelectorAndArgs,
+    uint256 value
+  ) internal returns (bool success, bytes memory data) {
+    // Load the system data
+    (address systemAddress, bool publicAccess) = Systems.get(resourceSelector);
+
+    // Check if the system exists
+    if (systemAddress == address(0)) revert IWorldErrors.ResourceNotFound(resourceSelector.toString());
+
+    // Allow access if the system is public or the caller has access to the namespace or name
+    if (!publicAccess) AccessControl.requireAccess(resourceSelector, caller);
+
+    // Call the system and forward any return data
+    (success, data) = resourceSelector.getNamespace() == ROOT_NAMESPACE // Use delegatecall for root systems (= registered in the root namespace)
+      ? WorldContextProvider.delegatecallWithContext({
+        msgSender: caller,
+        target: systemAddress,
+        funcSelectorAndArgs: funcSelectorAndArgs
+      })
+      : WorldContextProvider.callWithContext({
+        msgSender: caller,
+        target: systemAddress,
+        funcSelectorAndArgs: funcSelectorAndArgs,
+        value: value
+      });
+  }
+
+  /**
+   * Call the system at the given namespace and call any hooks registered for it.
+   * If the system is not public, the caller must have access to the namespace or name.
+   * Returns the success status and any return data.
+   */
+  function callWithHooks(
+    address caller,
+    bytes32 resourceSelector,
+    bytes memory funcSelectorAndArgs,
+    uint256 value
+  ) internal returns (bool success, bytes memory data) {
+    // Get system hooks
+    address[] memory hooks = SystemHooks.get(resourceSelector);
+
+    // Call onBeforeCallSystem hooks (before calling the system)
+    for (uint256 i; i < hooks.length; i++) {
+      ISystemHook hook = ISystemHook(hooks[i]);
+      hook.onBeforeCallSystem(caller, resourceSelector, funcSelectorAndArgs);
+    }
+
+    // Call the system and forward any return data
+    (success, data) = call(caller, resourceSelector, funcSelectorAndArgs, value);
+
+    // Call onAfterCallSystem hooks (after calling the system)
+    for (uint256 i; i < hooks.length; i++) {
+      ISystemHook hook = ISystemHook(hooks[i]);
+      hook.onAfterCallSystem(caller, resourceSelector, funcSelectorAndArgs);
+    }
+  }
+
+  /**
+   * Call the system at the given namespace and call any hooks registered for it.
+   * If the system is not public, the caller must have access to the namespace or name.
+   * Reverts with the error if the call was not successful.
+   * Else returns any return data.
+   */
+  function callWithHooksOrRevert(
```

Comment:
this is the previous behavior of `_call`. In addition we now have library functions that return a `success` flag instead of reverting, which is useful if we want to execute more logic if a call failed (eg. will be needed in #1364)